src/Authentication.Abstractions/AzureEnvironment.cs

```diff
@@ -25,9 +32,32 @@ namespace Microsoft.Azure.Commands.Common.Authentication.Abstractions
     /// </summary>
     [Serializable]
     public class AzureEnvironment : IAzureEnvironment
-    {
+    {        
+        private const int GetArmMetadataRetryCount = 3;
+        private const string ArmMetadataEnvVariable = "ARM_CLOUD_METADATA_URL";
+        private static readonly HttpClient Client = new HttpClient();        
+
         static IDictionary<string, AzureEnvironment> InitializeBuiltInEnvironments()
         {
+            IDictionary<string, AzureEnvironment> armAzureEnvironments = null;
+            try
+            {
+                var armMetadataRequestUri = Environment.GetEnvironmentVariable(ArmMetadataEnvVariable);
```

Comment:
The environment variable (ARM_CLOUD_METADATA_URL) is configured to point to the ARM metadata endpoint for a specific Azure cloud or Azure Stack. Azure PS/CLI will be initialized dynamically only if the variable is configured and if there are no errors retrieving/parsing the metadata; otherwise hard-coded values are used as a fallback